File: Week_04/cpp/src/_154_FindMinimumInRotatedSortedArrayII.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    int findMin(vector<int> &nums)
    {
        int l = 0, r = nums.size() - 1;
        while (l < r)
        {
            if (nums[r] > nums[l])
                break;
            int m = l + ((r - l) >> 1);
            if (nums[m] > nums[l])
                l = m + 1;
            else if (nums[m] < nums[l])
                r = m;
            else
                l++;
        }
        return nums[l];
    }
};

File: Week_04/kotlin/src/_154_FindMinimumInRotatedSortedArrayII.kt
class _154_BinarySearch_1 {
    fun findMin(nums: IntArray): Int {
        var l = 0
        var r = nums.lastIndex
        while (l < r) {
            val m = (l + r) ushr 1
            when {
                nums[m] > nums[r] -> {
                    l = m + 1
                }
                nums[m] < nums[r] -> {
                    r = m
                }
                else -> {
                    r--
                }
            }
        }
        return nums[l]
    }
}